app/controller/friend.js

```diff
@@ -0,0 +1,168 @@
+'use strict';
+
+const Controller = require('egg').Controller;
+class FriendController extends Controller {
+  async followNewUser() {
+    const { ctx } = this;
+    const userId = ctx.headers.userId;
+    const id = +ctx.params.id;
+
+    if (id === userId) {
+      throw new Error('Can not follow yourself');
+    }
+    const followerInfo = await ctx.service.user.getUserById(id);
+    if (!followerInfo) {
+      throw new Error('The Person is not exist');
+    }
+    const followRelation = await ctx.service.friend.checkFollowRelation(userId, id);
+    console.log('relation:', followRelation);
+    if (followRelation & (1 === 1)) {
+      throw new Error('you have already followed');
+    }
+    await ctx.service.friend.createFriend(userId, id);
+    return (ctx.body = {
+      status: 'success'
+    });
+  }
+
+  async removeFollower() {
+    const { ctx } = this;
+    const userId = ctx.headers.userId;
+    const id = +ctx.params.id;
+
+    const followRelation = await ctx.service.friend.checkFollowRelation(userId, id);
+    if (followRelation & (1 !== 1)) {
```

Comment:
Similar issue as L19. Actually I think `checkFollowRelation` is not a good idea, because the data might be changed after query due to concurrency request, if user send two follow requests at the same time, both of them will probably pass the check and result in two records in database.

Better solution:

- add unique index `(following, follower)`
- insert new relationship into friend table and catch `SequelizeUniqueConstraintError`
- use return value of `destroy` to check whether the relationship has been removed: https://sequelize.org/v7/class/src/model.js~Model.html#static-method-destroy

Reply:
Thanks forreviewing my homework. I hadn't realize so many bugs or shortcomings in it. And I had just committed for debugging.
- Using crypto instead of crypto-js.
- debug with plain text of pwd when updating user.
- debug with (followRelation & 1) === 1, that may caused by auto prettier :-(.

btw, I used for...in to get index of array, to add 'prefix' to args[0] and '\n' to args[args.length - 1], should i use for... instead?
for moment.js, it is a hlelpful package i used before, i will replace it since it is  in maintenance mode.
Then for multiple click of same request, i think unique index + transaction is a solution, but i didn't use it in this demo, i can do it this evening.

Reply:
> btw, I used for...in to get index of array, to add 'prefix' to args[0] and '\n' to args[args.length - 1], should i use for... instead?

If you only want to change the first and the last item, why not directly access to `args[0]` and `args[args.length-1]` without loop? If you do want to mutate array item, I think either `for (let i = 0; i < args.length; i++)` or `args.map()`is more readable.

Reply:
> If you only want to change the first and the last item, why not directly access to `args[0]` and `args[args.length-1]` without loop? If you do want to mutate each array item, I think either `for (let i = 0; i < args.length; i++)` or `args.map()`is more readable.

roger that, thanks